design/external-names.md

```diff
@@ -0,0 +1,47 @@
+# External Names Service
+
+_Status_: Draft
+
+This document looks to allow Contour to process Kubernetes service types `ExternalName` so that traffic can be proxied to dns hosts rather than only pod endpoints.
+
+## Goals
+
+- Allow for Contour to proxy traffic via Envoy to external dns endpoints
+- Utilize built-in Kubernetes object types (i.e. ExternalName service)
+
+## Non-goals
+
+- `externalIPs` will not be used, users who want to proxy to custom IP's should build their own service/endpoint object
+- Utilize a new CRD or ConfigMap style mapping
+
+## Background
+
+Contour currently watches Kubernetes services and endpoints, streaming them to via Envoy xDS CDS & EDS endpoints.
+Requests to envoy are routed directly to Kubernetes pods (i.e. endpoints).
+This proposal looks to implement Kubernetes service type `ExternalName` which allows for an `externalName` which is represented as a valid RFC-1123 hostname.
+
+## High-level Design
+
+Contour will need to watch for and process Kubernetes service types of `ExternalName` (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#servicespec-v1-core).
+When configuring EDS endpoints, for this service type Contour will pass the contents of the `externalName` field on the Kubernetes service to EDS instead of a PodIP.
+
+Additionally, users will need to configure ports on the service so Contour will understand how to proxy requests to the external endpoint.
```

Comment:
we're not really changing this; users still need to supply a Port variable to identify the DNS:PORT pair of the external service.

Reply:
I call it out because it's not required, I can create a service like this and k8s api is cool with it. I'll add a callout that we'll set the status on the IngressRoute object if a service is found but no ports.

```
apiVersion: v1
kind: Service
metadata:
  labels:
    run: nginx
  name: nginx
  namespace: default
spec:
  externalName: stevesloka.com
  type: ExternalName
status:
  loadBalancer: {}
  type: ExternalName
```

Reply:
good point. its probably worth stating it explicitly then -- the hosts come from `spec.externalName`, the Port comes from the `backend.servicePort` record.

We might also be able to use SRV records here, but that's out of scope for this proposal.